### PR TITLE
fix(backuptarget): backuptarget update skipped when enqueued by engineimage update

### PR DESCRIPTION
**Pull request by c3y1huang**
#### Which issue(s) this PR fixes:

Issue longhorn/longhorn#8045

#### What this PR does / why we need it:

Do not update the LastSyncedAt when the controller marks the BackupTarget unavailable.

### CODE REVIEW

File: controller/backup_target_controller.go [hunk 1]
```diff
@@ -332,17 +332,18 @@ func (btc *BackupTargetController) reconcile(name string) (err error) {
...
+	syncTime := &metav1.Time{}
 	existingBackupTarget := backupTarget.DeepCopy()
-	syncTime := metav1.Time{Time: time.Now().UTC()}
+
 	defer func() {
 		if err != nil {
 			return
 		}
-		if backupTargetClient != nil {
+		if backupTargetClient != nil && !syncTime.Time.IsZero() {
```
Comment (innobead) [inline controller/backup_target_controller.go:342]: It seems `syncTime` is a pointer, suggest doing a nil check still.

File: controller/backup_target_controller.go [hunk 2]
```diff
@@ -385,6 +386,8 @@ func (btc *BackupTargetController) reconcile(name string) (err error) {
 	}
 	defer engineClientProxy.Close()
 
+	syncTime = &metav1.Time{Time: time.Now().UTC()}
+
 	if err = btc.syncBackupVolume(backupTarget, backupTargetClient, syncTime, log); err != nil {
```
[the remark this answers is not in the document]
Comment (innobead) [reply controller/backup_target_controller.go:391]: Sounds good 👍 @c3y1huang

**Review by shuo-wu — DISMISSED** [review]
LGTM except for the comment Derek mentioned

**Review by innobead — APPROVED** [review]
LGTM.

**Comment by innobead** [comment]
@mergify backport v1.6.x v1.5.x

**Review by derekbit — APPROVED** [review]
LGTM

**Comment by c3y1huang** [comment]
Test result:
```
# ./run.sh -xs -k test_backuptarget_available_during_engine_image_not_ready --count=50
        if not 200 <= r.status <= 299:
>           raise ApiException(http_resp=r)
E           kubernetes.client.exceptions.ApiException: (422)
E           Reason: Unprocessable Entity
E           HTTP response headers: HTTPHeaderDict({'Audit-Id': 'bcfbd500-6056-4a81-a79c-42dc4d586083', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Kubernetes-Pf-Flowschema-Uid': '29bd88d4-8939-437a-abfe-8a43412279fe', 'X-Kubernetes-Pf-Prioritylevel-Uid': '9786bc1c-ed26-431d-a010-7e3bb4098f5b', 'Date': 'Tue, 27 Feb 2024 11:37:26 GMT', 'Content-Length': '187'})
E           HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server rejected our request due to an error in our request","reason":"Invalid","details":{},"code":422}

/usr/local/lib/python3.9/site-packages/kubernetes/client/rest.py:235: ApiException

= 1 failed, 50 passed, 18550 deselected, 1 warning in 3668.55s (1:01:08) =
```
Test at commit before https://github.com/longhorn/longhorn-manager/compare/bf0b1597cd311009183b4e45b1ba761d0c4d9703..88d3366bb4c062ff1e877a97cb1f7c43a73f44b8):
- test_backup* : https://ci.longhorn.io/job/private/job/longhorn-tests-regression/6537/
- test_backuptarget_available_during_engine_image_not_ready: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/6535/console

The failures are all similar, seems like a test case issue.

**Comment by c3y1huang** [comment]
> Test result:
> 
> ```
> # ./run.sh -xs -k test_backuptarget_available_during_engine_image_not_ready --count=50
> 
>                 # Scale up the engine image DaemonSet
>                 body = [{"op": "remove",
>                          "path": "/spec/template/spec/nodeSelector/foo"}]
> >               apps_api.patch_namespaced_daemon_set(
>                     name=ds_name, namespace='longhorn-system', body=body)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>         if not 200 <= r.status <= 299:
> >           raise ApiException(http_resp=r)
> E           kubernetes.client.exceptions.ApiException: (422)
> E           Reason: Unprocessable Entity
> E           HTTP response headers: HTTPHeaderDict({'Audit-Id': 'bcfbd500-6056-4a81-a79c-42dc4d586083', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Kubernetes-Pf-Flowschema-Uid': '29bd88d4-8939-437a-abfe-8a43412279fe', 'X-Kubernetes-Pf-Prioritylevel-Uid': '9786bc1c-ed26-431d-a010-7e3bb4098f5b', 'Date': 'Tue, 27 Feb 2024 11:37:26 GMT', 'Content-Length': '187'})
> E           HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server rejected our request due to an error in our request","reason":"Invalid","details":{},"code":422}
> 
> /usr/local/lib/python3.9/site-packages/kubernetes/client/rest.py:235: ApiException
> 
> = 1 failed, 50 passed, 18550 deselected, 1 warning in 3668.55s (1:01:08) =
> ```
> 
> Test at commit before https://github.com/longhorn/longhorn-manager/compare/bf0b1597cd311009183b4e45b1ba761d0c4d9703..88d3366bb4c062ff1e877a97cb1f7c43a73f44b8):
> 
> * test_backup* : https://ci.longhorn.io/job/private/job/longhorn-tests-regression/6537/
> * test_backuptarget_available_during_engine_image_not_ready: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/6535/console
> 
> The failures are all similar, seems like a test case issue.

Fix in https://github.com/longhorn/longhorn-tests/pull/1783.

**Pull request merged**